internal/engine/registry/invocation.go

```diff
@@ -2,6 +2,7 @@ package registry
 
```

Comment:
Is there documentation for the state of each invariant? Would probably be good to at least add high level docs on what each invariant does adjacent to the implementation of the invariant or within the docs somewhere.